src/rpc/http_server.cpp

```diff
@@ -0,0 +1,288 @@
+      return HTTP_BAD_REQUEST;
+    } catch (const std::exception& e) {
+      MWARNING("RPC request '/" << uri << "' request raised an exception: " << e.what());
+      return HTTP_ERROR;
+    } catch (...) {
+      MWARNING("RPC request '/" << uri << "' request raised an unknown exception");
+      return HTTP_ERROR;
+    }
+  }
+
+  http_response_code http_server::handle_json_rpc_request(
+    const epee::net_utils::http::http_request_info& query_info,
+    epee::net_utils::http::http_response_info& response_info,
+    connection_context& context)
+  {
+    auto& body = response_info.m_body;
+
+    rpc_request request;
+    request.context.admin = m_restricted;
+    request.body = jsonrpc_params{};
+    auto& epee_stuff = request.body.get_unchecked<jsonrpc_params>();
+    auto& ps = epee_stuff.first;
+    if(!ps.load_from_json(query_info.m_body))
+      return json_rpc_error(-32700, "Parse error", body);
+
+    epee::serialization::storage_entry id{std::string{}};
+    ps.get_value("id", id, nullptr);
+    std::string method;
+    if(!ps.get_value("method", method, nullptr))
+      return json_rpc_error(-32600, "Invalid Request", body);
+
+    auto it = rpc_commands.find(method);
+    if (it == rpc_commands.end() || it->second->is_binary)
+      return json_rpc_error(-32601, "Method not found", body);
+
+    const auto& command = *it->second;
+    if (m_restricted && !command.is_public)
+      return json_rpc_error(403, "Forbidden; this command is not available over public RPC", body);
+
+    constexpr std::array<lokimq::string_view, 3> json_rpc_ = {
+R"({"jsonrpc": "2.0", "id": )"_sv, R"(, "result": )"_sv, R"(
+}
+)"_sv};
+
+    std::string id_str;
+    {
+      std::ostringstream o;
+      epee::serialization::dump_as_json(o, id, 0 /*indent*/, false /*newlines*/);
+      id_str = o.str();
+    }
+
+    // Try to load "params" into a generic epee value; if it fails (because there is no "params")
+    // then we will replace it with an empty string to signal that no params were provided.
+    if (!ps.get_value("params", epee_stuff.second, nullptr))
+      request.body = ""_sv;
+
+    std::string result;
+    try {
+      result = command.invoke(std::move(request), m_server);
+    } catch (const parse_error& e) {
+      // This isn't really WARNable as it's the client fault; log at info level instead.
+      MINFO("JSON RPC request for '" << method << "' called with invalid data: " << e.what());
+      return json_rpc_error(-32602, "Invalid params", body);
+    } catch (const std::exception& e) {
+      MWARNING("json_rpc '" << method << "' request raised an exception: " << e.what());
+      return json_rpc_error(-32603, "Internal error", body);
+    } catch (...) {
+      MWARNING("json_rpc '" << method << "' request raised an unknown exception");
+      return json_rpc_error(-32603, "Internal error", body);
+    }
+
+    // The string is pre-serialized JSON.  But epee serialization is garbage so there's no way we
```

Comment:
> Isn't the string (result) post-serialized json already at this point?

It's only the serialized `result` value, not the full JSON RPC response value.  Previously this code was building a nested struct which basically embeds a JSON RPC response structure with the custom type embedded into that, which is gross.

However the string copying is also unpleasant.

> How about when we command.invoke, we pass in a string that already has the header attached, and when the response gets serialized it appends to that string (or writes into it), avoiding the copying stuff below except the response tail.

I came up with a better solution of just being able to multiple body strings to epee for it to concatenate when sending, which avoids the problem and keeps a cleaner interface (of just returning a string instead of needing to append it into a mutable parameter -- and even then not being able to properly reserve unless you violate separation of concerns by knowing that the caller is going to want to append a few more characters to it after you return).

> Can pre-allocate a string with thread_local std::string result(4096)

Using a static `thread_local` wouldn't work because the request body lifetime needs to be tied to the outgoing message, *not* the thread that created it: you want to build the response, say "send this", then go on to the next request without having to block on the send.  Epee sort of understands this by having the body be owned by the response object, but then screws it up by blocking the entire thread until the send is done anyway.  ZMQ in contrast takes ownership of the string and destroys it when the message is confirmed received by the remote, but the same thread might have processed 50 RPC commands before that happens.

Reply:
The new way looks good to me